File: rust/0001_twosum.rs
//https://leetcode.com/problems/two-sum/submissions/909162045/

use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut d: HashMap<i32, usize> = HashMap::with_capacity(nums.len());

        for (i, num) in nums.into_iter().enumerate() {
            let x = target - num;
            match(d.get(&x)) {
                Some(&index) => {
                    return vec![index as i32, i as i32];
                }
                None => {
                    d.insert(num, i);
                }
            }
        }
        return vec![];
    }
}

File: java/0001_two_sum.java
/*
 * https://leetcode.com/problems/two-sum/submissions/1278927980/
 */
import java.util.Map;
import java.util.HashMap;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<> ();
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i])) {
                return new int[] {map.get(nums[i]), i};
            } else {
                map.put(target - nums[i], i);
            }
        }
        return new int[0];
    }
}